keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1108 @@
+
+
+#### e2e tests
+
+Following scenarios need to be covered:
+
+
+
+* Cgroup settings when pod-level resources are set.
+* Validate scheduling and admission.
+* Validate the containers with no limits set are throttled on CPU when CPU usage reaches Pod level CPU limits.
+* Validate the containers with no limits set are OOMKilled when memory usage reaches Pod level memory limits.
+
+
+### Graduation Criteria
+
+
+#### Phase 1: Alpha (target 1.31)
+
+
+* Feature is disabled by default. It is an opt-in feature which can be enabled by setting the new fields in PodSpec.
+* Support the basic functionality for scheduler to consider pod-level resource requests to find a suitable node.
+* Support the basic functionality for kubelet to translate pod-level requests/limits to pod-level cgroup settings.
+* Unit test coverage.
+* E2E tests.
+* Documentation to announce the feature in alpha phase.
+
+
+#### Phase 2:  Beta (target 1.32)
+
+
+
+* User Feedback.
+* Feature is disabled by default. It is an opt-in feature which can be enabled by setting the new fields in PodSpec.
+* Support for all features that are scoped out of alpha phase in the KEP design.
+* Unit test coverage.
+* E2E tests.
+* Documentation update to announce feature in beta phase.
+* Benchmark tests for resource usage.
+
+
+#### GA (stable)
+
+
+* TBD
+
+
+### Upgrade / Downgrade Strategy
+
+The existing workloads will not have any impact since they won't be using the new field `resources` in PodSpec 
+at pod-level. If a cluster is upgraded, the cluster administrator will experience pre-upgrade behavior until
+`resources` field is specified at pod-level. 
+
+If the workloads are not using the new feature, downgrade won't be affected. However, if it is required to
+downgrade a cluster that has workloads using the pod-level limits functionality, following steps need to be
+performed:
+
+* Translate pod-level resource specifications to container-level specifications.
+
+* Delete existing workloads and recreate them with container-level specifications.
+
+* Drain the nodes one by one to prevent the new pods from being scheduled on them.
+
+* Downgrade the kubelet on the drained nodes to a version that doesn't support pod-level resource specification
+fields.
+
+* Downgrade the control plane components.
+
+* Uncordon the nodes, making it available for scheduling again.
+
+
+### Version Skew Strategy
+
+Pod-level resource specification is an opt-in feature that gets enabled by supplying the `resources` stanza 
+in the `spec` at pod level. For this feature to work correctly, it must be enabled in all parts of the 
+cluster (scheduler, API server, kubelet). However, if a node is running an older version of kubelet that 
+doesn't support this feature, it will ignore the pod-level resoucr specifications and continue to use the 
+aggregated container-level values (backward compatible). Applications must be prepared for this version skew 
```

Comment:
maybe during the transition time we should require users specify container limits in the apiserver until all supported kubelets support this field, to prevent the risk of a user not specifying any limits and messing up scheduling/runtime of the application

Reply:
For some users, they'd like to specify pod level resources so that they don't have to deal with individual container resources, so this'd defeat the purpose for them. 

This also seems to be a common issue we'd face for similar features. Maybe we can follow the convention?

Reply:
yeah totally, and I think that's a good end-goal, but requiring them in the short term to specify container limits despite the pod limits being in place would ensure their pods are in the right QOS class and are limited/schedule correctly